Implement the `entryMode` extension

aurangzaibumer opened the pull request:
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Issue# [1272](https://github.com/google/android-fhir/issues/1272)

**Description**
Clear and concise code change description. 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature

**Screenshots/Video**

https://user-images.githubusercontent.com/35099184/170076662-ca522aa3-6ce5-48ed-96bc-85a65ec0938e.mp4



**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

aurangzaibumer commented:
@dubdabasoduba as per discussion on standup, there is a PR for submit button https://github.com/google/android-fhir/pull/1215 this needs to be merged first in order to work on the submit button so probably we can create a separate ticket for submit button implementation so this can be merged if reviewed.

joiskash reviewed (changes requested):
Thinking about this a little more, I figured that this more than just validation.. This is triggering some logic on page change events within the QuestionnaireFragment. So we could refactor as such. This will allow more complex use cases such as :

1. Not allowing to switch page until a minimum amount of time has passed
2. Not allowing page change if the answers are exactly the same as the previous questionnaire answered . (this can happen in cases were there is a monetary incentive to complete forms)
 
Validation of all the items on the page can be the default implementation provided by the library. Developers can choose to add this in their Strategies too.

By having a context, the main point is that the strategy can be changed during run time.  So a developer can do something like 

`dataConfig.questionnairePageChangeStrategy.setStrategy(newStrategy)`

at any point in the code during run time. 

Create a context class that will be stored in the DataCapture Config
```
class QuestionnairePageChangeEventContext{
  private var pageChangeStrategy : PageChangeStrategy 

  fun setStrategy(pageChangeStrategy : PageChangeStrategy ){
      this.pageChangeStrategy = pageChangeStrategy
  }
  
  //This function will be called in the QuestionnaireViewModel goToNextPage and if returns true  then pageFlow.value will be changed.

  fun pageNextEvent( list: List<QuestionnaireItemViewItem> ):Boolean{
  return pageChangeStrategy.shouldGoToNextPage(list)
  }

//This function will be called in the QuestionnaireViewModel goToPreviousPage and if returns true  then pageFlow.value will be changed.

  fun pagePreviousEvent( list: List<QuestionnaireItemViewItem> ):Boolean{
  return pageChangeStrategy.shouldGoToPreviousPage(list)
  }
}
```
The strategy interface will be like this. 

```
interface PageChangeStrategy {

  fun shouldGoToPreviousPage(list: List<QuestionnaireItemViewItem>): Boolean

  fun shouldGoToNextPage(list: List<QuestionnaireItemViewItem>): Boolean

}
```

The default implementation can be like :

```
class DefaultPageChangeStrategy: PageChangeStrategy{

  override fun shouldGoToPreviousPage(list: List<QuestionnaireItemViewItem>): Boolean{
    return list.any { it.isErrorTriggered }
  }

  override fun shouldGoToNextPage(list: List<QuestionnaireItemViewItem>): Boolean
    return list.any { it.isErrorTriggered }
  }
```

and finally the change in the viewmodel can be as such:

'''
internal class QuestionnaireViewModel(application: Application, state: SavedStateHandle) :{
  private val questionnairePageEventContext by lazy {
    DataCapture.getConfiguration(getApplication()).questionnairePageEventContext ?: this
  }
.
.
.
  internal fun goToPreviousPage() {
    if (questionnairePageEventContext.pagePreviousEvent()){
              pageFlow.value = pageFlow.value!!.previousPage()
     }
  }
  
  internal fun goToNextPage() {
    if (questionnairePageEventContext.pageNextEvent()){
              pageFlow.value = pageFlow.value!!.nextPage()
     }
  }
}

'''

aurangzaibumer commented:
> Thinking about this a little more, I figured that this more than just validation.. This is triggering some logic on page change events within the QuestionnaireFragment. So we could refactor as such. This will allow more complex use cases such as :
> 
> 1. Not allowing to switch page until a minimum amount of time has passed
> 2. Not allowing page change if the answers are exactly the same as the previous questionnaire answered . (this can happen in cases were there is a monetary incentive to complete forms)
> 
> Validation of all the items on the page can be the default implementation provided by the library. Developers can choose to add this in their Strategies too.
> 
> By having a context, the main point is that the strategy can be changed during run time. So a developer can do something like
> 
> `dataConfig.questionnairePageEventContext.setStrategy(newStrategy)`
> 
> at any point in the code during run time.
> 
> Create a context class that will be stored in the DataCapture Config
> 
> ```
> class QuestionnairePageChangeEventContext{
>   private var pageChangeStrategy : PageChangeStrategy 
> 
>   fun setStrategy(pageChangeStrategy : PageChangeStrategy ){
>       this.pageChangeStrategy = pageChangeStrategy
>   }
>   
>   //This function will be called in the QuestionnaireViewModel goToNextPage and if returns true  then pageFlow.value will be changed.
> 
>   fun pageNextEvent( list: List<QuestionnaireItemViewItem> ):Boolean{
>   return pageChangeStrategy.shouldGoToNextPage(list)
>   }
> 
>   //This function will be called in the QuestionnaireViewModel goToPreviousPage and if returns true  then pageFlow.value will be changed.
> 
>   fun pagePreviousEvent( list: List<QuestionnaireItemViewItem> ):Boolean{
>   return pageChangeStrategy.shouldGoToPreviousPage(list)
>   }
> }
> ```
> 
> The strategy interface will be like this.
> 
> ```
> interface PageChangeStrategy {
> 
>   fun shouldGoToPreviousPage(list: List<QuestionnaireItemViewItem>): Boolean
> 
>   fun shouldGoToNextPage(list: List<QuestionnaireItemViewItem>): Boolean
> 
> }
> ```
> 
> The default implementation can be like :
> 
> ```
> class DefaultPageChangeStrategy: PageChangeStrategy{
> 
>   override fun shouldGoToPreviousPage(list: List<QuestionnaireItemViewItem>): Boolean{
>     return list.any { it.isErrorTriggered }
>   }
> 
>   override fun shouldGoToNextPage(list: List<QuestionnaireItemViewItem>): Boolean
>     return list.any { it.isErrorTriggered }
>   }
> ```
> 
> and finally the change in the viewmodel can be as such:
> 
> ```
> internal class QuestionnaireViewModel(application: Application, state: SavedStateHandle) :{
>   private val questionnairePageEventContext by lazy {
>     DataCapture.getConfiguration(getApplication()).questionnairePageEventContext // declare and initialize this in the DataConfig. It should never be null
>   }
> .
> .
> .
>   internal fun goToPreviousPage() {
>     if (questionnairePageEventContext.pagePreviousEvent()){
>               pageFlow.value = pageFlow.value!!.previousPage()
>      }
>   }
>   
>   internal fun goToNextPage() {
>     if (questionnairePageEventContext.pageNextEvent()){
>               pageFlow.value = pageFlow.value!!.nextPage()
>      }
>   }
> }
> ```
> 
> Some one from google can comment on this, @jingtang @kevinmost

feedback implemented, please review @joiskash . Thanks

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -40,9 +40,14 @@ import org.hl7.fhir.r4.model.ValueSet
 
 internal class QuestionnaireViewModel(application: Application, state: SavedStateHandle) :
-  AndroidViewModel(application) {
+  AndroidViewModel(application), QuestionnaireValidation {
```
QuestionnaireValidation is not needed anymore right? I think it can be removed

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -371,6 +382,19 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
     }
   }
+
+  fun holdCurrentItemState(items: List<QuestionnaireItemViewItem>) {
```
Can this be made internal?

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -371,6 +382,19 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+    currentPageItems = items
+  }
+
+  override fun isPageNotValidated(list: List<QuestionnaireItemViewItem>): Boolean {
+    return list.any { it.isErrorTriggered }
+  }
+
+  override val shouldCheckValidationOnSubmit = true
+
+  override val shouldCheckValidationOnSwitchingPages = true
```
Remove this?

aurangzaibumer replied:
resolved. thanks

joiskash reviewed (changes requested):
Great work, left a few comments. I request someone from google and Ona to review this before you write the UTs. @jingtang10 @f-odhiambo

aurangzaibumer commented on datacapture/src/main/java/com/google/android/fhir/datacapture/DefaultPageChangeStrategy.kt:
```diff
@@ -0,0 +1,30 @@
+package com.google.android.fhir.datacapture
+
+import com.google.android.fhir.datacapture.views.QuestionnaireItemViewItem
+
```
done. thanks

jingtang10 commented:
@santosh-pingle can you leave some comments please

aurangzaibumer commented:
> Great work, left a few comments. I request someone from google and Ona to review this before you write the UTs. @jingtang10 @f-odhiambo

@dubdabasoduba

santosh-pingle commented:
> @santosh-pingle can you leave some comments please

sure.

santosh-pingle commented on datacapture/src/main/java/com/google/android/fhir/datacapture/DefaultPageChangeStrategy.kt:
```diff
@@ -0,0 +1,33 @@
+class DefaultPageChangeStrategy : PageChangeStrategy {
+
+  override fun shouldGoToPreviousPage(list: List<QuestionnaireItemViewItem>): Boolean {
+    return !list.any { it.isErrorTriggered }
```
I see duplication of the code in `inshouldGoToPreviousPage()` and  `shouldGoToNextPage()`.  
Avoid duplication of code : 
Approach 1 : Introduce one more api `isErrorOnCurrentPage(list: List<QuestionnaireItemViewItem>): Boolean` in 'PageChangeStrategy'  and move error validation logic here, and call this api from above two api.

Approach 2 : or just create private api `isErrorOnCurrentPage(list: List<QuestionnaireItemViewItem>): Boolean` here instead of introducing new api in interface.

aurangzaibumer replied:
Used approached 2. Thanks for highlighting. Can you please review.

santosh-pingle commented on datacapture/src/main/java/com/google/android/fhir/datacapture/DefaultPageChangeStrategy.kt:
```diff
@@ -0,0 +1,33 @@
+  }
+
+  override fun shouldGoToNextPage(list: List<QuestionnaireItemViewItem>): Boolean {
+    return !list.any { it.isErrorTriggered }
```
please refer earlier comment.

santosh-pingle commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnairePageChangeEventContext.kt:
```diff
@@ -0,0 +1,45 @@
+
+import com.google.android.fhir.datacapture.views.QuestionnaireItemViewItem
+
+class QuestionnairePageChangeEventContext {
```
can we please add api doc?

santosh-pingle commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnairePageChangeEventContext.kt:
```diff
@@ -0,0 +1,45 @@
+    this.pageChangeStrategy = pageChangeStrategy
+  }
+
+  // This function will be called in the QuestionnaireViewModel goToNextPage and if returns true
```
Can we please use api doc comment format
/* **/

santosh-pingle commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnairePageChangeEventContext.kt:
```diff
@@ -0,0 +1,45 @@
+    return pageChangeStrategy.shouldGoToNextPage(list)
+  }
+
+  // This function will be called in the QuestionnaireViewModel goToPreviousPage and if returns true
```
ditto

santosh-pingle commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -43,6 +43,11 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
   AndroidViewModel(application) {
   /** The current questionnaire as questions are being answered. */
   internal val questionnaire: Questionnaire
+  lateinit var currentPageItems: List<QuestionnaireItemViewItem>
```
can we declare it as private if it is only used here?

santosh-pingle reviewed (commented):
left few minor comments.
Also please check whether we need `demo/src/main/assets/new-patient-registration-paginated.json` file change?

aurangzaibumer commented:
> left few minor comments. Also please check whether we need `demo/src/main/assets/new-patient-registration-paginated.json` file change?

was checking the questionnaire for pagination. I'll revert it back. Thanks.

aurangzaibumer commented:
Can you please review it again @santosh-pingle I had resolved the changes asked. Thanks

aurangzaibumer commented:
> Can you please review it again @santosh-pingle I had resolved the changes asked. Thanks

cc @dubdabasoduba

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/DefaultPageChangeStrategy.kt:
```diff
@@ -0,0 +1,41 @@
+import com.google.android.fhir.datacapture.views.QuestionnaireItemViewItem
+
+/*
+ * This Page strategy restricts user on every page when required fields are missing,
+ * implements PageChangeStrategy to give support for switching pages of the Questionnaire
```
```suggestion
 * Only allows pagination when there are no validation errors on the current page.
```

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/DefaultPageChangeStrategy.kt:
```diff
@@ -0,0 +1,41 @@
+   * isErrorTriggered will be true if any required field has not yet been provided with an answer
+   * */
+  private fun isErrorOnCurrentPage(list: List<QuestionnaireItemViewItem>): Boolean {
+    return !list.any { it.isErrorTriggered }
```
```suggestion
    return list.none { it.isErrorTriggered }
```

jingtang10 commented:
can you rename the PR: `Allow/disallow pagination based on validation result (e.g. can't go to next page with validation error)`

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -382,6 +393,11 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
     }
   }
+
+  internal fun holdCurrentItemState(items: List<QuestionnaireItemViewItem>) {
```
so the consumer of this view model will use this function to set the items after getting the items from this view model. that's a roundabout way of caching the items.

we could just cache these items when they are being returned in the `getQuestionnaireState` function?

aurangzaibumer replied:
I used the hard coded integer for pagination since I did know my exact page for now to check the current page items but it wasn't updated as per current page (isErrorTriggered were all false in those items that I received from below snippet although I had few errors showing on my page).

```
getQuestionnaireState(
      questionnaireItemList = questionnaire.item,
      questionnaireResponseItemList = questionnaireResponse.item,
      pagination = QuestionnairePagination(1, 0),
    ).items 
```

Can you please correct me if I'm wrong. Thanks

aurangzaibumer commented on datacapture/src/main/java/com/google/android/fhir/datacapture/DataCaptureConfig.kt:
```diff
@@ -45,7 +45,14 @@ data class DataCaptureConfig(
+   * Holds the current strategy that will be followed during page change events. This can be
+   * customized on runtime by setting the customized strategy
+   */
+  var questionnairePageEventContext: QuestionnairePageChangeEventContext =
```
`DateCaptureConfig is global. Is this better as a per-questionnaire setting?`

Since We have to give developers the right to update the PageChangeStrategy that is why I added QuestionnairePageChangeEventContext in DataCaptureConfig. If I understood your question correctly.

jingtang10 replied:
my question is why is this not a setting for individual questionnaires.

aurangzaibumer commented:
> * the getQuestionnaireState method should probably incorporate questionnaire validation - at the moment it doesn't do any of that, it's probably incorrect. this also address @kevinmost 's todo he left in that function. 

This is done, please review and share feedback


> * it's possible as a result of the previous item, we might not need to directly call validation code in the answer changed callback. all that needs to be done is to get a new questionnaire state. i'm not sure if this is true so please do test it out.

I have tried to remove the validation from onBind and onAnswerChanged but the sequence of the method calling is different as I expected. whenever you give any input, **onAnswerChanged** gets called first, after that **getQuestionnaireState** gets called hence we cannot use **validationResult** from the **questionnaireViewItem** model in the Method **onAnswerChanged**

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -392,10 +440,7 @@ internal data class QuestionnaireState(
-  val currentPageIndex: Int,
-  val lastPageIndex: Int,
-)
+internal data class QuestionnairePagination(val currentPageIndex: Int, val lastPageIndex: Int)
```
we discussed this, and we agreed to add more params here: can go to next page and can go to previous page. no?

@joiskash

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/DefaultPageChangeStrategy.kt:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Only allows pagination when all required fields are validated on the current page .
+ */
+class DefaultPageChangeStrategy : PageChangeStrategy {
```
we discussed this and decided to remove page change startegy.

aurangzaibumer replied:
Sorry, can you please re-explain this because I don't really remember removing this strategy. thanks

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -328,9 +361,24 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
     return QuestionnaireState(items = items, pagination = pagination)
   }
 
+  private fun canSwitchPages() {
+    viewModelScope.launch {
+      _paginatedNextButtonStateFlow.emit(
+        questionnairePageEventContext.pageNextEvent(currentPageItems)
+      )
+      _paginatedPreviousButtonStateFlow.emit(
+        questionnairePageEventContext.pagePreviousEvent(currentPageItems)
+      )
+    }
+  }
```
the approach we agreed to take is that we add more states in the questionnarie state, so that it includes whether it's ok to go to previous/next page, so the UI will use that information to enable/disable the paging buttons

joiskash replied:
Well I see that @shelaghm has described why it might not be such a good idea to enable / disable the buttons .

https://github.com/google/android-fhir/issues/1272

@jingtang10

jingtang10 replied:
@aurangzaibumer i replied in the issue - i think i am persuaded that we shouldn't have disabld buttons. what this means is that we need to do validation in `getQuestionnarieState` and include the validation errors in the questionnarie item view items.

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/DefaultPageChangeStrategy.kt:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Only allows pagination when all required fields are validated on the current page .
+ */
+class DefaultPageChangeStrategy : PageChangeStrategy {
```
Can this be made private?

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnairePageChangeEventContext.kt:
```diff
@@ -0,0 +1,65 @@
+ * Responsible for holding current Page strategy that can changed/set dynamically on runtime by the
+ * Questionnaire
+ */
+class QuestionnairePageChangeEventContext {
```
Same as above

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnairePageChangeEventContext.kt:
```diff
@@ -0,0 +1,65 @@
+/*
+ * Responsible giving support methods for Questionnaire Page switching
+ *  */
+interface PageChangeStrategy {
```
Same as above

aurangzaibumer replied:
made it internal

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -160,7 +182,15 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+    } else {
+      //WIP highlight errors if any
+      viewModelScope.launch {
+        currentPageItems.forEach { it.modified = true }
```
Not sure why we are doing this.. seems hacky..can you please remind me?  Why do we mark all items as modified? Is this because of the change that we made to avoid validation messages being displayed as soon as the page is open?

aurangzaibumer replied:
This is WIP, actually I was trying to refresh the current page items because I do have the latest validationResult for each items, whenever next button is pressed, I need to show the validation where applicable I was trying to call the items because they were not being modified by the user. Currently this is WIP I'm working things around. thanks

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -160,7 +182,15 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+      //WIP highlight errors if any
+      viewModelScope.launch {
+        currentPageItems.forEach { it.modified = true }
+        _refresh.emit(currentPageItems)
```
Looks like we have created a new flow for page change.. is this a workaround to https://github.com/google/android-fhir/issues/1427 ?

aurangzaibumer replied:
this will be emited when next page is pressed. I'm trying to refresh the list when next button is pressed (in case if current page is not validated, the errors will be thrown to the required field where applicable that is why I added modified true). Yes its a bit hacky, please suggest any alternative. thanks

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -411,3 +449,7 @@ internal fun QuestionnairePagination.nextPage(): QuestionnairePagination {
   return copy(currentPageIndex = currentPageIndex + 1)
 }
+
+internal fun QuestionnairePagination.refresh() : QuestionnairePagination {
```
Is this being used anywhere?

aurangzaibumer replied:
No, this is currently WIP. let me remove it.

joiskash reviewed (changes requested):
Sorry about the delay, left a few comments !

Tarun-Bhardwaj commented:
@aurangzaibumer , is there an ETA you could provide so that @joiskash can continue with the code review?
CC: @f-odhiambo

aurangzaibumer commented:
> @aurangzaibumer , is there an ETA you could provide so that @joiskash can continue with the code review? CC: @f-odhiambo

Discussed the issues/blockers with JING & Kashyap today in the SDK team, Requested a discussion meeting by tomorrow or next week (on Jing's availability).

aurangzaibumer commented:
As per discussion with Jing

1. He suggested to separate the Validation Changes from the Pagination changes so I'm removing the validation changes from this PR. This PR will only reflect changes that is related to **_pagination_**
2. The second thing He suggested is to use the ENUM approach instead of Page Change Strategy

Tarun-Bhardwaj commented:
@aurangzaibumer @joiskash , is this PR blocked? Any support you need here to move forward?

aurangzaibumer commented:
Made changes related to enum approach
For now keeping
- the validationResult code base
- Paging change strategy
will remove once discussion is done with @jingtang10 today

As per discussion below PR needs to be merged first in order to consume the validationResult code base from the QuestionnaireViewModel
https://github.com/google/android-fhir/pull/1468

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -156,11 +174,37 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+        if (currentPageItems.none { !it.validationResult!!.isValid }) {
+          pageFlow.value = pageFlow.value!!.nextPage()
+        } else {
+          //show toast that validation is failing
```
Same comment as above

joiskash commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -411,3 +463,9 @@ internal fun QuestionnairePagination.nextPage(): QuestionnairePagination {
   return copy(currentPageIndex = currentPageIndex + 1)
 }
+
+enum class EntryMode(val value: Int) {
```
Can you please add documentation that is borrowed from http://build.fhir.org/ig/HL7/sdc/ValueSet-entryMode.html and http://build.fhir.org/ig/HL7/sdc/StructureDefinition-sdc-questionnaire-entryMode.html

joiskash reviewed (changes requested):
Good work! Left a few comments

Tarun-Bhardwaj commented:
@aurangzaibumer , could you please provide an ETA for addressing review comments from @joiskash ?

aurangzaibumer commented:
> @aurangzaibumer , could you please provide an ETA for addressing review comments from @joiskash ?

This current PR is dependent on https://github.com/google/android-fhir/pull/1468 because we will be using the validation logic from 1468 PR and remove it from here. Me and Jing are currently working on #1468 so we can close that ASAP

Tarun-Bhardwaj commented:
@aurangzaibumer , considering that #1468 was merged last week, can the pending bits be closed on this PR as well?
CC: @f-odhiambo @jingtang10

aurangzaibumer commented:
> Reference

I'm facing an issue, for suppose when first try to open the questionnaire there are no errors shown and when user taps on next page (on restrict mode), the application should be showing errors on required fields (if not entered) in this case we have to re-validate the questionnaire (even if they are not being modified by the user) and also have to reflect the errors (where applicable) on the UI. Now what I did was that I iterate over currentPageItems so each element could  have latest validationResult (because there might be a case where fields are not being modified by the user yet) and then I have to update the UI in-order to reflect the errors

`//Revalidating it because there's a chance if they the fields are not being modified yet by the User
       currentPageItems.forEach {
         it.validationResult?.isValid = QuestionnaireResponseItemValidator.validate(
           it.questionnaireItem,
           it.answers,
           this@QuestionnaireViewModel.getApplication()
         ).isValid

       }`
       
can you please share your opinion if that's the right strategy. @jingtang10 

cc @maimoonak @joiskash

jingtang10 commented:
> > @aurangzaibumer , considering that #1468 was merged last week, can the pending bits be closed on this PR as well? CC: @f-odhiambo @jingtang10
> 
> I'm facing an issue, for suppose when we open the questionnaire there are no errors shown and when user taps on next page (on restrict mode), the application should be showing errors on required fields (if not entered) in this case we have to re-validate the questionnaire (even if they are not being modified by the user) and also have to reflect the errors (where applicable) on the UI. Now what I did was that I iterate over **currentPageItems** so each element could have latest validationResult (because there might be a case where fields are not being modified by the user yet) and then I have to update the UI in-order to reflect the errors
> 
> ```
>  currentPageItems.forEach {
>          it.validationResult?.isValid = QuestionnaireResponseItemValidator.validate(
>            it.questionnaireItem,
>            it.answers,
>            this@QuestionnaireViewModel.getApplication()
>          ).isValid
> 
>        }
> ```
> 
> can you please share your opinion if that's the right strategy. @jingtang10
> 
> cc @maimoonak @joiskash

can you add all the items to the modified set when the user tries to go to the next page?

aurangzaibumer commented:
> > > @aurangzaibumer , considering that #1468 was merged last week, can the pending bits be closed on this PR as well? CC: @f-odhiambo @jingtang10
> > 
> > 
> > I'm facing an issue, for suppose when we open the questionnaire there are no errors shown and when user taps on next page (on restrict mode), the application should be showing errors on required fields (if not entered) in this case we have to re-validate the questionnaire (even if they are not being modified by the user) and also have to reflect the errors (where applicable) on the UI. Now what I did was that I iterate over **currentPageItems** so each element could have latest validationResult (because there might be a case where fields are not being modified by the user yet) and then I have to update the UI in-order to reflect the errors
> > ```
> >  currentPageItems.forEach {
> >          it.validationResult?.isValid = QuestionnaireResponseItemValidator.validate(
> >            it.questionnaireItem,
> >            it.answers,
> >            this@QuestionnaireViewModel.getApplication()
> >          ).isValid
> > 
> >        }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > can you please share your opinion if that's the right strategy. @jingtang10
> > cc @maimoonak @joiskash
> 
> can you add all the items to the modified set when the user tries to go to the next page?

  I have used a boolean that will be set as true whenever next page is pressed as 

`isPaginationButtonPressed = true
        modificationCount.update { it + 1 }`
        
 and updated a check on questionnareState as 
  
  `modifiedQuestionnaireResponseItemSet.contains(questionnaireResponseItem) || isPaginationButtonPressed`
  
   and it will be reset to false before pageFlow value changes as 
   
   `if (currentPageItems.none { !it.validationResult!!.isValid }) {
          isPaginationButtonPressed = false
          pageFlow.value = pageFlow.value!!.nextPage()
        }`
        
        
        

https://user-images.githubusercontent.com/35099184/180257590-020eeeb3-ec79-4333-9c56-9ce2a968aef0.mp4



can you please review if it's a better strategy @jingtang10 

cc @maimoonak @f-odhiambo

jingtang10 commented:
Thanks umer - this is much better

aurangzaibumer commented:
> Thanks umer - this is much better

thanks @jingtang10 , I'll be update the documentation and the test cases based on this implementation.

cc @f-odhiambo @maimoonak

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/DefaultPageChangeStrategy.kt:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Only allows pagination when all required fields are validated on the current page .
+ */
+internal class DefaultPageChangeStrategy : PageChangeStrategy {
```
we have discussed this - can you remove this interface?

jingtang10 replied:
we have said that we'll be using entry mode http://build.fhir.org/ig/HL7/sdc/ValueSet-entryMode.html which is generalised to the page level

jingtang10 commented on datacapture/src/main/res/values/strings.xml:
```diff
@@ -30,4 +30,7 @@
     <string name="select_date">Select date</string>
     <string name="select_time">Select time</string>
     <string name="hyphen">-</string>
+    <string
+        name="message_fill_required_fields"
+    >Please fill the required fields first</string>
```
delete this

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:
```diff
@@ -160,6 +166,8 @@ open class QuestionnaireFragment : Fragment() {
 
     const val SUBMIT_REQUEST_KEY = "submit-request-key"
+
+    const val EXTRA_QUESTIONNAIRE_PAGING_STRATEGY = "questionnaire-paging-strategy"
```
remove this

aurangzaibumer commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -225,11 +240,32 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
   internal fun goToPreviousPage() {
-    pageFlow.value = pageFlow.value!!.previousPage()
+    when (entryMode) {
+      EntryMode.PRIOR_EDIT, EntryMode.RANDOM -> {
```
according to doc
prior-edit`: display new top level page or question only when previous page or question answered. can go back.

so that is why I gave **prior-edit** and **random** the same functionality when tapping on **previous page**

jingtang10 commented:
please revert the license header changes - makes it very difficult to review the code.

thanks!

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -76,9 +78,22 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
       }
   }
 
+  private val entryMode: EntryMode by lazy {
+    EntryMode.valueOf(
+      questionnaire
+        .extension
+        .firstOrNull { it.url == EXTENSION_ENTRY_MODE_URL }
+        ?.value
+        ?.toString()
+        ?.uppercase(Locale.getDefault())
+        ?: EntryMode.RANDOM.value.uppercase(Locale.getDefault())
+    )
+  }
```
you need to create a new class called `MoreQuestionnaires.kt` and put the logic to load entry mode there. also add tests.

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -225,11 +240,32 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+          pageFlow.value = pageFlow.value!!.nextPage()
+        }
+      }
+      EntryMode.RANDOM, EntryMode.SEQUENTIAL -> {
```
i think sequential should be together with prior-edit?

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -225,11 +240,32 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
-    pageFlow.value = pageFlow.value!!.nextPage()
+    when (entryMode) {
+      EntryMode.PRIOR_EDIT -> {
+
```
remove random blank line

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -76,9 +78,22 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+        ?.value
+        ?.toString()
+        ?.uppercase(Locale.getDefault())
+        ?: EntryMode.RANDOM.value.uppercase(Locale.getDefault())
```
by the way i think this field should be nullable. don't assume here that random is the default.

aurangzaibumer replied:
Okay but for suppose if there is no value (null) for entryMode and user taps on the pagination button (next or previous), in that case what should happen? thanks

jingtang10 replied:
I think probably prior edit? But keep that logic in the view model instead of here.

jingtang10 replied:
I know this is the view model file. But you need to move this to a separate file just as utility. And the view model will check things and decide the default behaviour

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -527,3 +568,9 @@ internal fun QuestionnairePagination.nextPage(): QuestionnairePagination {
   return copy(currentPageIndex = currentPageIndex + 1)
 }
+
+enum class EntryMode(val value: String) {
+  PRIOR_EDIT("prior_edit"),
+  RANDOM("random"),
+  SEQUENTIAL("sequential")
```
move this to `MoreQuestionnaires.kt`

jingtang10 commented on datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:
```diff
@@ -1167,6 +1167,108 @@ class QuestionnaireViewModelTest(
+      .isEqualTo(QuestionnairePagination(currentPageIndex = 0, lastPageIndex = 1))
+    assertThat(state.items).hasSize(2)
+    viewModel.goToNextPage()
+    assertTrue(viewModel.questionnairePageEventContext.pageNextEvent(state.items))
```
do these tests even pass - there's no page event context any more?

aurangzaibumer replied:
I haven't updated the unit test cases yet. I did only develop new implementation so that you can review and if you agree to that then i will be updating the unit test cases.

jingtang10 replied:
Sure that's fine. I think you can probably start writing tests.

maimoonak commented on datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaires.kt:
```diff
@@ -38,3 +38,21 @@ val Questionnaire.targetStructureMap: String?
+    val entryMode =
+      this.extension.firstOrNull { it.url == EXTENSION_ENTRY_MODE_URL }?.value?.toString()
+    return when (entryMode?.lowercase()) {
+      "prior-edit" -> EntryMode.PRIOR_EDIT
```
can we add these values to enum as code? just as we do have for other enums for FHIR?

maimoonak commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -225,11 +229,31 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+        pageFlow.value = pageFlow.value!!.previousPage()
+      }
+      EntryMode.SEQUENTIAL -> {
+        // Previous questions and submitted answers cannot be viewed or edited.
```
is it possible to show a toast here? OR use else and log it as warning.

aurangzaibumer commented:
All comments are resolved. Can you please review and share you feedback @jingtang10 
Thanks

jingtang10 commented on datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:
```diff
@@ -1167,6 +1168,186 @@ class QuestionnaireViewModelTest(
   }
 
+  @Test
+  fun shouldAllowUserToSwitchBetweenPages_entryMode_prior_edit() = runBlocking {
```
can you use the new kotlin back quote format: https://kotlinlang.org/docs/coding-conventions.html#names-for-test-methods

i think you should probably also have separate test cases in each entry mode for going prev/next.

aurangzaibumer replied:
I have added three unit test cases for each entry mode

aurangzaibumer commented on datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaires.kt:
```diff
@@ -38,3 +38,27 @@ val Questionnaire.targetStructureMap: String?
+internal const val EXTENSION_ENTRY_MODE_URL: String =
+  "http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-entryMode"
+
+val Questionnaire.entryMode: EntryMode
```
I have added three unit test case for each entry mode

joiskash commented on datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:
```diff
@@ -1167,6 +1168,195 @@ class QuestionnaireViewModelTest(
   }
 
+  @Test
+  fun `should Allow user to move forward And Back Using Prior Entry Mode`() = runBlocking {
```
The case in this looks statement looks weird.

joiskash reviewed (changes requested):
Wow this PR looks so clean! Great work, just a small comment, we should probably add a test that checks the default entry mode if there is no extension for entry mode in the questionnaire.

aurangzaibumer commented:
> Wow this PR looks so clean! Great work, just a small comment, we should probably add a test that checks the default entry mode if there is no extension for entry mode in the questionnaire.

added a case when no entryMode is defined in the questionnaire. Thanks

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -225,11 +233,31 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+        isPaginationButtonPressed = true
+        modificationCount.update { it + 1 }
+
+        if (currentPageItems.none { !it.validationResult!!.isValid }) {
```
```suggestion
        if (currentPageItems.all { it.validationResult!!.isValid }) {
```

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -76,9 +78,13 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
   /** The current questionnaire response as questions are being answered. */
   private val questionnaireResponse: QuestionnaireResponse
 
+  private var isPaginationButtonPressed = false
```
add a comment

jingtang10 commented on datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:
```diff
@@ -1167,6 +1168,252 @@ class QuestionnaireViewModelTest(
   }
 
+  @Test
+  fun `should allow user to move forward and back using prior entry-mode`() = runBlocking {
```
separate all forward and backward cases?

jingtang10 commented on datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:
```diff
@@ -1167,6 +1168,252 @@ class QuestionnaireViewModelTest(
+      }
+    val viewModel = createQuestionnaireViewModel(questionnaire)
+    val state = viewModel.questionnaireStateFlow.first()
+    assertThat(state.pagination)
+      .isEqualTo(QuestionnairePagination(currentPageIndex = 0, lastPageIndex = 1))
+    assertThat(state.items).hasSize(2)
```
don't have to do theses

jingtang10 commented on datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:
```diff
@@ -1167,6 +1168,252 @@ class QuestionnaireViewModelTest(
   }
 
+  @Test
+  fun `should allow user to move forward and back using prior entry-mode`() = runBlocking {
```
missing a test cases which is you can't go forward without answering the questions in prior edit mode.

aurangzaibumer replied:
added the unit tests for prior-mode and sequential

jingtang10 reviewed (approved):
thanksk @aurangzaibumer

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -231,19 +243,42 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+        isPaginationButtonPressed = true
+        modificationCount.update { it + 1 }
+
+        if (currentPageItems.all { it.validationResult!!.isValid }) {
```
```suggestion
        if (currentPageItems.all { it.validationResult.isValid }) {
```

jingtang10 replied:
this needs to be changed with the validation result code change

aurangzaibumer replied:
thanks for the suggestion. resolved

jingtang10 replied:
actually i've realised we have an issue here - if the user hasn't touched any question, does that not make the this check pass? which means that the user can always go to the next page without editing anything in the current page?

can you test this? thanks

i think this logic probably should be changed to checking if everything is valid.

aurangzaibumer replied:
yes you are right although it was working fine having both valid and not validated check because when we click on paginated buttons, it validates the items and receive validation either (invalid or valid). I have updated the logic though.

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -411,35 +448,39 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+              questionnaireResponseItem.answer.first().item
+            },
+        )
+    // holding updated items state
```
remove this comment.

aurangzaibumer commented:
@jingtang10 can you please review on your availability. thanks

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -231,19 +244,42 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
-    currentPageIndexFlow.value = nextPageIndex
+    when (entryMode) {
+      EntryMode.PRIOR_EDIT, EntryMode.SEQUENTIAL -> {
+        isPaginationButtonPressed = true
+        modificationCount.update { it + 1 }
```
can you add some comments to explain the necessity of having these lines here?

jingtang10 replied:
```suggestion
        // Force update validation results for all questions on the current page
        isPaginationButtonPressed = true
        modificationCount.update { it + 1 }
```

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -231,19 +244,43 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
-    currentPageIndexFlow.value = nextPageIndex
+    when (entryMode) {
+      EntryMode.PRIOR_EDIT, EntryMode.SEQUENTIAL -> {
+        // Force update validation results for all questions on the current page
+        isPaginationButtonPressed = true
+        modificationCount.update { it + 1 }
```
do you think this makes sense? i think we can save some additional validation if the user has already pressed the button.

```suggestion
        if (!isPaginationButtonPressed) {
          // Force update validation results for all questions on the current page
          isPaginationButtonPressed = true
          modificationCount.update { it + 1 }
        }
```

aurangzaibumer replied:
If i could map a case 
1. User opened the questionnaire and taps on the next button without touching any items
2. isPaginated is FALSE, it will trigger the validation for the current page items 
3. now errors are shown the required fields, now user tried to tap again on the next button 
4. since isPaginated is already TRUE, this will save the additional validation in this _case_. 
You are right @jingtang10 . This makes sense to me.

cc @maimoonak

jingtang10 commented on datacapture/src/test/java/com/google/android/fhir/datacapture/MoreQuestionnairesTest.kt:
```diff
@@ -79,4 +80,25 @@ class MoreQuestionnairesTest {
     assertThat(questionnaire.isPaginated).isFalse()
   }
+
+  @Test
```
as @joiskash suggested, add a test case here for null case.

aurangzaibumer replied:
case added in MoreQuestionnaireTest class. thanks
also there was a test added `should allow user to move forward when no entry-mode is defined` in QuestionnaireViewModelTest as well.

jingtang10 replied:
that's great - thanks!

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -231,19 +244,45 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+    when (entryMode) {
+      EntryMode.PRIOR_EDIT, EntryMode.SEQUENTIAL -> {
+        if (!isPaginationButtonPressed) {
+          // Force update validation results for all questions on the current page
```
```suggestion
          // Force update validation results for all questions on the current page. This is needed when the user has not answered any questions so no validation has been done.
```

jingtang10 commented on datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:
```diff
@@ -156,11 +174,37 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
+        if (currentPageItems.none { !it.validationResult!!.isValid }) {
+          pageFlow.value = pageFlow.value!!.nextPage()
+        } else {
+          //show toast that validation is failing
```
Let's keep this in a follow-up pr.

santosh-pingle commented:
@aurangzaibumer can you please add some questionnaire examples?

aurangzaibumer commented:
> @aurangzaibumer can you please add some questionnaire examples?

```
"extension": [ 
    {
      "url" : "http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-entryMode",
      "valueCode" : "prior-edit"
    } 
 ]
```
for more entry-mode codes : http://build.fhir.org/ig/HL7/sdc/ValueSet-entryMode.html
